## Задача№2

Загружаем датасеты digits и breast_cancer из sklearn.datasets

In [12]:
from sklearn import cross_validation, datasets, metrics, neighbors 
from sklearn.datasets import load_breast_cancer, load_digits
from sklearn.model_selection import cross_val_score

import sklearn.naive_bayes as nb
from sklearn.naive_bayes import GaussianNB, BernoulliNB, MultinomialNB

/projects/sage/sage-7.5/local/lib/python2.7/site-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [13]:
digits = datasets.load_digits()
breast_cancer = datasets.load_breast_cancer()

In [14]:
digits.data[:3]

array([[  0.,   0.,   5.,  13.,   9.,   1.,   0.,   0.,   0.,   0.,  13.,
         15.,  10.,  15.,   5.,   0.,   0.,   3.,  15.,   2.,   0.,  11.,
          8.,   0.,   0.,   4.,  12.,   0.,   0.,   8.,   8.,   0.,   0.,
          5.,   8.,   0.,   0.,   9.,   8.,   0.,   0.,   4.,  11.,   0.,
          1.,  12.,   7.,   0.,   0.,   2.,  14.,   5.,  10.,  12.,   0.,
          0.,   0.,   0.,   6.,  13.,  10.,   0.,   0.,   0.],
       [  0.,   0.,   0.,  12.,  13.,   5.,   0.,   0.,   0.,   0.,   0.,
         11.,  16.,   9.,   0.,   0.,   0.,   0.,   3.,  15.,  16.,   6.,
          0.,   0.,   0.,   7.,  15.,  16.,  16.,   2.,   0.,   0.,   0.,
          0.,   1.,  16.,  16.,   3.,   0.,   0.,   0.,   0.,   1.,  16.,
         16.,   6.,   0.,   0.,   0.,   0.,   1.,  16.,  16.,   6.,   0.,
          0.,   0.,   0.,   0.,  11.,  16.,  10.,   0.,   0.],
       [  0.,   0.,   0.,   4.,  15.,  12.,   0.,   0.,   0.,   0.,   3.,
         16.,  15.,  14.,   0.,   0.,   0.,   0.,   8.,  13.

In [15]:
breast_cancer.data[:3]

array([[  1.79900000e+01,   1.03800000e+01,   1.22800000e+02,
          1.00100000e+03,   1.18400000e-01,   2.77600000e-01,
          3.00100000e-01,   1.47100000e-01,   2.41900000e-01,
          7.87100000e-02,   1.09500000e+00,   9.05300000e-01,
          8.58900000e+00,   1.53400000e+02,   6.39900000e-03,
          4.90400000e-02,   5.37300000e-02,   1.58700000e-02,
          3.00300000e-02,   6.19300000e-03,   2.53800000e+01,
          1.73300000e+01,   1.84600000e+02,   2.01900000e+03,
          1.62200000e-01,   6.65600000e-01,   7.11900000e-01,
          2.65400000e-01,   4.60100000e-01,   1.18900000e-01],
       [  2.05700000e+01,   1.77700000e+01,   1.32900000e+02,
          1.32600000e+03,   8.47400000e-02,   7.86400000e-02,
          8.69000000e-02,   7.01700000e-02,   1.81200000e-01,
          5.66700000e-02,   5.43500000e-01,   7.33900000e-01,
          3.39800000e+00,   7.40800000e+01,   5.22500000e-03,
          1.30800000e-02,   1.86000000e-02,   1.34000000e-02,
       

In [16]:
print ("target names digits: {}".format(digits.target_names))

target names digits: [0 1 2 3 4 5 6 7 8 9]


In [17]:
print ("target names breast_cancer: {}".format(breast_cancer.target_names))

target names breast_cancer: ['malignant' 'benign']


In [25]:
print 'breast_cancer:'

print 'Multinomial', cross_val_score(nb.MultinomialNB(), breast_cancer.data, breast_cancer.target, cv=5).mean()
print 'Gaussian', cross_val_score(nb.GaussianNB(), breast_cancer.data, breast_cancer.target, cv=5).mean()
print 'Bernoulli', cross_val_score(nb.BernoulliNB(), breast_cancer.data, breast_cancer.target, cv=5).mean()

breast_cancer:
Multinomial 0.89637552905
Gaussian 0.940377068103
Bernoulli 0.627425933051


In [26]:
print 'digits:'

print 'Multinomial', cross_val_score(nb.MultinomialNB(), digits.data, digits.target, cv=5).mean()
print 'Gaussian', cross_val_score(nb.GaussianNB(), digits.data, digits.target, cv=5).mean()
print 'Bernoulli', cross_val_score(nb.BernoulliNB(), digits.data, digits.target, cv=5).mean()

digits:
Multinomial 0.871470302548
Gaussian 0.806520755555
Bernoulli 0.82477104598


Максимальное качество для digits: на Multinomial 0.871470302548
для breast-cancer: Gaussian 0.940377068103

Верно только лишь (d)